File: CodeGenDotNet 159059-1372 (a)/VBCode/Component Shells/WinSupport/Utility.vb
' ====================================================================
'  Copyright ©2004, Kathleen Dollard, All Rights Reserved.
' ====================================================================
'   I'm distributing this code so you'll be able to use it to see code
'   generation in action and I hope it will be useful and you'll enjoy 
'   using it. This code is provided "AS IS" without warranty, either 
'   expressed or implied, including implied warranties of merchantability 
'   and/or fitness for a particular purpose. 
'  ====================================================================
'  Summary: Home for utility methods

Option Strict On
Option Explicit On 
Imports System


Public Class Utility

   Public Const border As Int32 = 6
   Public Const margin As Int32 = 4
   Public Const vbcrlf As String = Microsoft.VisualBasic.vbCrLf

   Public Shared Function SubstringBefore( _
                  ByVal s As String, _
                  ByVal sBefore As String) _
                  As String
      Dim iPos As Int32 = s.IndexOf(sBefore)
      If iPos = -1 Then
         Return s
      Else
         Return s.Substring(0, iPos)
      End If
   End Function

   Public Shared Function ParentForm( _
                  ByVal ctl As Windows.Forms.Control) _
                  As Windows.Forms.Form
      Do While Not TypeOf ctl Is Windows.Forms.Form
         ctl = ctl.Parent
      Loop
      If Not ctl Is Nothing Then
         Return CType(ctl, Windows.Forms.Form)
      End If
   End Function

   Public Shared Function ParentForm( _
                  ByVal ctl As Windows.Forms.MenuItem) _
                  As Windows.Forms.Form
      Return ctl.Parent.GetMainMenu.GetForm
   End Function

   Public Shared Function ParentForm( _
                  ByVal sender As Object) _
                  As Windows.Forms.Form
      If TypeOf sender Is Windows.Forms.MenuItem Then
         Return Utility.ParentForm(CType(sender, Windows.Forms.MenuItem))
      Else
         Return Utility.ParentForm(CType(sender, Windows.Forms.Control))
      End If
   End Function


   Public Shared Function Position( _
                  ByVal graphics As Drawing.Graphics, _
                  ByVal ctl As Windows.Forms.Control, _
                  ByVal iTop As Int32) _
                  As Int32
      Const c As Char = "K"c
      Dim sizef As Drawing.SizeF
      Dim sizefText As Drawing.SizeF
      Dim sizefSpace As Drawing.SizeF
      Dim sizeFNew As Drawing.SizeF
      Dim chars As Int32
      Dim cnt As Int32
      Dim noResize As Boolean
      Dim multiline As Boolean
      Dim iLines As Int32
      Dim iChars As Int32

      ctl.Top = iTop

      If TypeOf ctl Is Windows.Forms.DataGrid Then
         noResize = True
      ElseIf TypeOf ctl Is Windows.Forms.TextBox Then
         chars = CType(ctl, Windows.Forms.TextBox).MaxLength
         multiline = CType(ctl, Windows.Forms.TextBox).Multiline
      ElseIf TypeOf ctl Is Windows.Forms.ComboBox Then
         chars = CType(ctl, Windows.Forms.ComboBox).MaxLength
      End If

      If chars = 0 Then
         cnt = 1
      Else
         cnt = chars
      End If

      If Not noResize Then
         sizefText = graphics.MeasureString(ctl.Text, ctl.Font)
         sizef = graphics.MeasureString(New String(c, cnt), ctl.Font)
         sizefSpace = New Drawing.SizeF(ctl.Parent.Size.Width - ctl.Left - border, _
                        ctl.Parent.Size.Height - ctl.Top - border)

         If sizefText.Width > sizef.Width Then
            sizef.Width = sizefText.Width
         End If

         If chars > 32000 Then
            sizeFNew = New Drawing.SizeF(ctl.Parent.Width - ctl.Left, sizef.Height + 6)
         Else
            sizeFNew = sizef
         End If

         If multiline Then
            sizeFNew = New Drawing.SizeF(sizeFNew.Width, sizefSpace.Height)
         End If

         ctl.Size = New Drawing.Size(CInt(sizeFNew.Width), CInt(sizeFNew.Height))
      End If
      iTop = ctl.Bottom + margin
      Return iTop

   End Function

   Public Shared Function Position( _
                     ByVal graphics As Drawing.Graphics, _
                     ByVal lbl As Windows.Forms.Label, _
                     ByVal ctl As Windows.Forms.Control, _
                     ByVal labelWidth As Int32, _
                     ByVal iTop As Int32) _
                     As Int32
      Dim ret As Int32
      lbl.Width = labelWidth
      ctl.Left = lbl.Right + margin
      ret = Position(graphics, ctl, iTop)
      Position(graphics, lbl, iTop)
      Return ret
   End Function

   Public Shared Function IsEmpty(ByVal val As Object) As Boolean
      If TypeOf val Is Int16 Then
      ElseIf TypeOf val Is Int16 Then
         Return CType(val, Int16) = 0
      ElseIf TypeOf val Is Int32 Then
         Return CType(val, Int32) = 0
      ElseIf TypeOf val Is Int64 Then
         Return CType(val, Int64) = 0
      ElseIf TypeOf val Is Double Then
         Return CType(val, Double) = 0
      ElseIf TypeOf val Is Single Then
         Return CType(val, Single) = 0
      ElseIf TypeOf val Is String Then
         Return CType(val, String) = ""
      ElseIf TypeOf val Is Guid Then
         Return Guid.op_Equality(CType(val, Guid), Guid.Empty)
      ElseIf TypeOf val Is DateTime Then
         Return CType(val, DateTime) = #12:00:00 AM#
      ElseIf TypeOf val Is Byte Then
      End If
   End Function


   Public Shared Sub BindField( _
               ByVal control As Windows.forms.Control, _
               ByVal propertyName As String, _
               ByVal dataSource As Object, _
               ByVal dataMember As String)

      Dim bd As Windows.Forms.Binding
      Dim index As Integer

      For index = control.DataBindings.Count - 1 To 0 Step -1
         bd = control.DataBindings.Item(index)
         If bd.PropertyName = propertyName Then
            control.DataBindings.Remove(bd)
         End If
      Next
      control.DataBindings.Add(propertyName, dataSource, dataMember)

   End Sub

   Public Shared Sub BindEvents( _
               ByVal control As Windows.Forms.Control, _
               ByVal controlProperty As String, _
               ByVal source As Object, _
               ByVal sourceProperty As String, _
               ByVal delegFormat As Windows.Forms.ConvertEventHandler, _
               ByVal delegParse As Windows.Forms.ConvertEventHandler)

      Dim binding As Windows.Forms.Binding = control.DataBindings(controlProperty)

      If Not binding Is Nothing Then
         control.DataBindings.Remove(binding)
      End If

      binding = New Windows.Forms.Binding(controlProperty, source, sourceProperty)

      If Not delegFormat Is Nothing Then
         AddHandler binding.Format, delegFormat
      End If

      If Not delegParse Is Nothing Then
         AddHandler binding.Parse, delegParse
      End If

      control.DataBindings.Add(binding)

   End Sub

   Public Shared Function InvokeSharedMethod( _
            ByVal type As System.Type, _
            ByVal method As String, _
            ByVal ParamArray params() As Object) _
            As Object
      Do While type.GetMember(method).GetLength(0) = 0
         type = type.BaseType
         If type Is GetType(Object) Then
            type = Nothing
            Exit Do
         End If
      Loop
      Return type.InvokeMember(method, _
               Reflection.BindingFlags.InvokeMethod, _
               Nothing, Nothing, params)
   End Function

   Public Shared Function InvokeInstanceMethod( _
               ByVal type As System.Type, _
               ByVal method As String, _
               ByVal obj As Object, _
               ByVal ParamArray params() As Object) _
               As Object
      Return type.InvokeMember(method, _
               Reflection.BindingFlags.IgnoreCase Or Reflection.BindingFlags.Instance Or Reflection.BindingFlags.InvokeMethod, _
               Nothing, obj, params)
   End Function

   Public Shared Function InvokeSharedPropertyGet( _
               ByVal type As System.Type, _
               ByVal method As String, _
               ByVal ParamArray params() As Object) _
               As Object
      ' Static properties are a bit tricky to retrieve. 
      ' This is an explicit walk of the hierarchy looking 
      ' for the method, assuming if its found the signature
      ' is right. KAD 1/22/04
      Do While type.GetMember(method).GetLength(0) = 0
         type = type.BaseType
         If type Is GetType(Object) Then
            type = Nothing
            Exit Do
         End If
      Loop
      Return type.InvokeMember(method, _
               Reflection.BindingFlags.GetProperty, _
               Nothing, Nothing, params)
   End Function

   Public Shared Function CreateInstance( _
               ByVal type As System.Type, _
               ByVal ParamArray params() As Object) _
               As Object
      Return Activator.CreateInstance(type, params)
   End Function

   Public Shared Function GetWidth( _
               ByVal stringToMeasure As String, _
               ByVal font As Drawing.Font, _
               ByVal graphics As Drawing.Graphics) As Integer
      Return CInt(graphics.MeasureString(stringToMeasure, font).Width)
   End Function

End Class

File: CodeGenDotNet 159059-1372 (a)/CSharp/Component Shells/WinSupport/Utility.cs
// ====================================================================
//  Copyright ©2004, Kathleen Dollard, All Rights Reserved.
// ====================================================================
//   I'm distributing this code so you'll be able to use it to see code
//   generation in action and I hope it will be useful and you'll enjoy 
//   using it. This code is provided "AS IS" without warranty, either 
//   expressed or implied, including implied warranties of merchantability 
//   and/or fitness for a particular purpose. 
//  ====================================================================
//  Summary: Home for utility methods

using System;

namespace KADGen.WinSupport
{
	public class Utility
	{
		public const int border = 6;
		public const int margin = 4;
		public const string vbcrlf = Microsoft.VisualBasic.ControlChars.CrLf;

		public static string SubstringBefore(
			string s,
			string sBefore)
		{
			int iPos = s.IndexOf(sBefore);
			if( iPos == -1 )
			{
				return s;
			}
			else
			{
				return s.Substring(0 , iPos);
			}
		}

		public static System.Windows.Forms.Form ParentForm(
			System.Windows.Forms.Control ctl)
		{
			while( ! (ctl is System.Windows.Forms.Form ))
			{
				ctl = ctl.Parent;
			}
			if( ctl != null )
			{
				return ( System.Windows.Forms.Form ) ctl;
			}
			else
			{
				return null;
			}
		}

		public static System.Windows.Forms.Form ParentForm(
			System.Windows.Forms.MenuItem ctl)
		{
			return ctl.Parent.GetMainMenu().GetForm();
		}

		public static  System.Windows.Forms.Form ParentForm(
			object sender)
		{     
			if( sender is System.Windows.Forms.MenuItem )
			{
				return Utility.ParentForm((System.Windows.Forms.MenuItem) sender );
			}
			else
			{
				return Utility.ParentForm(( System.Windows.Forms.Control ) sender);
			}
		}


		public static int Position(
			System.Drawing.Graphics graphics,
			System.Windows.Forms.Control ctl,
			int iTop)
		{         
			const char c  = 'K';
			System.Drawing.SizeF sizef;
			System.Drawing.SizeF sizefText;
			System.Drawing.SizeF sizefSpace;
			System.Drawing.SizeF sizeFNew;
			int chars = 0;
			int cnt = 0;
			bool noResize = false;
			bool multiline = false;
			int iLines = 0;
			int iChars = 0;

			ctl.Top = iTop;

			if( ctl is System.Windows.Forms.DataGrid )
			{
				noResize = true;
			}
			else if ( ctl is System.Windows.Forms.TextBox )
			{
				chars = (( System.Windows.Forms.TextBox ) ctl).MaxLength;
				multiline = (( System.Windows.Forms.TextBox ) ctl).Multiline;
			}
			else if ( ctl is System.Windows.Forms.ComboBox )
			{
				chars = (( System.Windows.Forms.ComboBox ) ctl).MaxLength;
			}

			if( chars == 0 )
			{
				cnt = 1;
			}
			else
			{
				cnt = chars;
			}

			if( ! noResize )
			{
				sizefText = graphics.MeasureString(ctl.Text , ctl.Font);
				sizef = graphics.MeasureString(new string(c , cnt) , ctl.Font);
				sizefSpace = new System.Drawing.SizeF(ctl.Parent.Size.Width - ctl.Left - border ,
					ctl.Parent.Size.Height - ctl.Top - border);

				if( sizefText.Width > sizef.Width )
				{
					sizef.Width = sizefText.Width;
				}

				if( chars > 32000 )
				{
					sizeFNew = new System.Drawing.SizeF(ctl.Parent.Width - ctl.Left , sizef.Height + 6);
				}
				else
				{
					sizeFNew = sizef;
				}

				if( multiline )
				{
					sizeFNew = new System.Drawing.SizeF(sizeFNew.Width , sizefSpace.Height);
				}

				ctl.Size = new System.Drawing.Size(System.Convert.ToInt32( sizeFNew.Width ) , System.Convert.ToInt32( sizeFNew.Height ));
			}
			iTop = ctl.Bottom + margin;
			return iTop;

		}

		public static int Position(
			System.Drawing.Graphics graphics,
			System.Windows.Forms.Label lbl,
			System.Windows.Forms.Control ctl,
			int labelWidth,
			int iTop)
		{          
			int ret ;
			lbl.Width = labelWidth;
			ctl.Left = lbl.Right + margin;
			ret = Position(graphics , ctl , iTop);
			Position(graphics , lbl , iTop);
			return ret;
		}

		public static bool IsEmpty( object val ) 
		{
			if ( val is short )
			{
				return (short)val  == 0;
			}
			else if ( val is int )
			{
				return (int)val == 0;
			}
			else if ( val is long )
			{
				return (long)val  == 0;
			}
			else if  ( val is double )
			{
				return (double)val  == 0;
			}
			else if ( val is float )
			{
				return (float)val  == 0;
			}
			else if ( val is string )
			{
				return (string) val == "";
			}
			else if ( val is Guid )
			{
				return ((Guid) val == Guid.Empty);
			}
			else if ( val is DateTime )
			{
				return ( DateTime ) val  == System.DateTime.MinValue;
			}
			else if ( val is Byte )
			{
				return ((Byte) val == 0);
			}
			else
			{
				return false;
			}
		}


		public static void BindField(
			System.Windows.Forms.Control control,
			string propertyName,
			object dataSource,
			string dataMember)
		{
			System.Windows.Forms.Binding bd ;

			for( int index = control.DataBindings.Count - 1 ; index >= 0 ; index-- )
			{
				bd = control.DataBindings[index];
				if( bd.PropertyName == propertyName )
				{
					control.DataBindings.Remove(bd);
				}
			}
			control.DataBindings.Add(propertyName , dataSource , dataMember);
		}

		public static void BindEvents(
			System.Windows.Forms.Control control,
			string controlProperty,
			object source,
			string sourceProperty,
			System.Windows.Forms.ConvertEventHandler delegFormat,
			System.Windows.Forms.ConvertEventHandler delegParse)
		{
			System.Windows.Forms.Binding binding = control.DataBindings[controlProperty];

			if( binding != null )
			{
				control.DataBindings.Remove(binding);
			}

			binding = new System.Windows.Forms.Binding(controlProperty , source , sourceProperty);

			if( delegFormat != null )
			{
				binding.Format += new System.Windows.Forms.ConvertEventHandler(delegFormat);
			}

			if( delegParse != null )
			{
				binding.Parse += new System.Windows.Forms.ConvertEventHandler(delegParse);
			}

			control.DataBindings.Add(binding);

		}

		public static object InvokeSharedMethod(
			System.Type type,
			string method,
			params object[] parms)
		{
			while( type.GetMember(method).GetLength(0) == 0 )
			{
				type = type.BaseType;
				if( type == typeof(object) )
				{
					type = null;
					break;
				}
			}
			return type.InvokeMember(method ,
				System.Reflection.BindingFlags.InvokeMethod ,
				null , null , parms);
		}

		public static object InvokeInstanceMethod(
			System.Type type,
			string method,
			object obj,
			params object[] parms)
		{
			return type.InvokeMember(method ,
				System.Reflection.BindingFlags.IgnoreCase | System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.InvokeMethod ,
				null , obj , parms);
		}

		public static object InvokeSharedPropertyGet(
			System.Type type,
			string method,
			params object[] parms)
			// Static properties are a bit tricky to retrieve. 
			// This is an explicit walk of the hierarchy looking 
			// for the method, assuming if its found the signature
			// is right. KAD 1/22/04
	{
			while( type.GetMember(method).GetLength(0) == 0 )
			{
				type = type.BaseType;
				if( type == typeof(Object) )
				{
					type = null;
					break;
				}
			}
			return type.InvokeMember(method ,
				System.Reflection.BindingFlags.GetProperty ,
				null , null , parms);
		}

		public static object CreateInstance(
			System.Type type,
			params object[] parms)
		{
			return Activator.CreateInstance(type , parms);
		}

		public static int GetWidth(
			string stringToMeasure,
			System.Drawing.Font font,
			System.Drawing.Graphics graphics) 
		{
			return System.Convert.ToInt32(graphics.MeasureString(stringToMeasure , font).Width);
		}

	}
}